### PR TITLE
dev adamw graph config

### PR DESCRIPTION
**Pull request by MARD1NO**

### CODE REVIEW

File: python/oneflow/nn/optimizer/adamw.py [hunk 1]
```diff
@@ -141,3 +141,72 @@ def step(self, closure: Callable = None):
...
+
+            train_conf.set_loss_scale_factor(scale)
+            optimizer_conf.set_base_learning_rate(lr)
+            
+            optimizer_conf.mutable_adam_conf().set_beta1(beta1)
+            optimizer_conf.mutable_adam_conf().set_beta2(beta2)
+            optimizer_conf.mutable_adam_conf().set_epsilon(epsilon)
+            optimizer_conf.mutable_adam_conf().set_do_bias_correction(
+               False
+            )  # TODO(zzk): Check this option
```
Comment (MARD1NO) [inline python/oneflow/nn/optimizer/adamw.py:205]: 这部分的weight decay需要这么处理？

lazy下是传入了一系列列表，可以指定哪些参数有weight decay，哪些参数没有
代码见：https://github.com/Oneflow-Inc/oneflow/blob/master/python/oneflow/compatible/single_client/ops/optimizer.py#L1513-L1532 

但是看了下eager没有对应的接口， 我直接这里set进去就可以了？

Comment (strint) [reply python/oneflow/nn/optimizer/adamw.py:205]: 这里先和eager对齐，如果eager没有，就写TODO注明下。

Comment (wyg1997) [reply python/oneflow/nn/optimizer/adamw.py:205]: https://github.com/Oneflow-Inc/oneflow/blob/master/python/oneflow/nn/optimizer/sgd.py#L155

对参数单独设置，是不是参考这里对l2的配置就可以了

Comment (MARD1NO) [reply python/oneflow/nn/optimizer/adamw.py:205]: https://pytorch.org/docs/master/generated/torch.optim.AdamW.html#torch.optim.AdamW 我看torch这边也没这个指定参数有weightdecay的功能。那还是感觉和eager对齐就可以了？我直接set weight decay给全部变量

Comment (strint) [reply python/oneflow/nn/optimizer/adamw.py:205]: 另外，所谓的部分weight_decay，在eager是有的，就是
```
l2 = param_group["weight_decay"]
```
不同参数不同配置，是通过param_group实现的。

Comment (wyg1997) [reply python/oneflow/nn/optimizer/adamw.py:205]: torch是可以的

```
adamw = optim.AdamW([{"param": params1, "weight_decay": 0.1}, {"param": param2, "weight_decay": 0.2}])
```

Comment (strint) [reply python/oneflow/nn/optimizer/adamw.py:205]: torch的weight_decay，默认是我们variable.l2，只有adamw不是

Comment (wyg1997) [reply python/oneflow/nn/optimizer/adamw.py:205]: 哦哦，你们在说这个，那它没有这个参数的话我们可以不设置

Comment (MARD1NO) [reply python/oneflow/nn/optimizer/adamw.py:205]: 明白了，感谢

File: python/oneflow/test/graph/test_graph_adamw_optim.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+            return (param, v, s)
+
+        for i in range(train_iters):
+            (x, vt, st) = np_train_one_iter(random_grad_seq[i])
+            np_res_list.append(x)
+
+    train_by_numpy()
+
+    test_case.assertTrue(
+        np.allclose(np_res_list, of_res_list, rtol=0.0001, atol=0.0001)
```
Comment (Ldpe2G) [inline python/oneflow/test/graph/test_graph_adamw_optim.py:116]: allclose 能直接比较两个 list of numpy 数组吗

File: python/oneflow/nn/optimizer/adamw.py [hunk 3]
```diff
@@ -141,3 +137,36 @@ def step(self, closure: Callable = None):
...
+            optimizer_conf = train_conf.mutable_optimizer_conf().Add()
+            lr = param_group["lr"]
+            weight_decay = param_group["weight_decay"]
+            beta1 = param_group["betas"][0]
+            beta2 = param_group["betas"][1]
+            epsilon = param_group["eps"]
+
+            # TODO(): optimizer_conf need to have loss_scale_factor field to support multi scale factor
+            base_scale = train_conf.loss_scale_factor()
+            assert math.isclose(base_scale, 1, rel_tol=1e-4), "nn.Graph only support one scale factor at the moment, base_scale {} vs scale {}".format(
```
Comment (Ldpe2G) [inline python/oneflow/nn/optimizer/adamw.py:152]: 这里还有必要对 base_scale 做限制吗，我看https://github.com/Oneflow-Inc/oneflow/pull/5821 这个pr里去掉了

File: python/oneflow/test/graph/test_graph_adamw_optim.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    )
+
+
+@flow.unittest.skip_unless_1n1d()
+class TestAdamW(flow.unittest.TestCase):
+    def test_adamw1(test_case):
+        compare_with_numpy_adamw(
```
Comment (Ldpe2G) [inline python/oneflow/test/graph/test_graph_adamw_optim.py:120]: 这里改成 多组参数组合

File: python/oneflow/nn/optimizer/adamw.py (hunk 3 above)
Comment (strint) [reply python/oneflow/nn/optimizer/adamw.py:152]: 的确没有必要了

Comment (strint) [reply python/oneflow/nn/optimizer/adamw.py:152]: 同步一下那次讨论的结论
- loss scale是amp的功能之一，逻辑是防止loss过小，浮点截断后，导致回传梯度消失，所以先增大loss * scale，之后再grad/scale
- loss scale不属于optimizer的参数，所以optimizer中的scale在5821中都清理掉了
- optimizer调用modle_update_op的那个scale参数，是scale weight的一个接口，给内部调整weight用的通用接口

File: python/oneflow/nn/optimizer/adamw.py [hunk 5]
```diff
@@ -141,3 +137,36 @@ def step(self, closure: Callable = None):
...
+            optimizer_conf.mutable_adam_conf().set_epsilon(epsilon)
+            optimizer_conf.mutable_adam_conf().set_do_bias_correction(
+                False
+            )  # TODO(zzk): Check this option
+
+            optimizer_conf.mutable_weight_decay_conf().set_weight_decay_rate(
+                weight_decay
+            )
+            for param in param_group.parameters:
+                vars_conf[param].weight_decay = weight_decay
```
Comment (strint) [inline python/oneflow/nn/optimizer/adamw.py:169]: vars_conf[param].weight_decay = weight_decay
这个需要去掉，没有作用

File: python/oneflow/nn/optimizer/adamw.py [hunk 6]
```diff
@@ -141,3 +137,36 @@ def step(self, closure: Callable = None):
...
+                if not param.requires_grad:
+                    continue
```
Comment (strint) [inline python/oneflow/nn/optimizer/adamw.py:171]: 这里也调整下，我之前这样写，是后面之前有比较长的代码，现在后面很短，就没必要了

File: python/oneflow/nn/optimizer/adamw.py [hunk 7]
```diff
@@ -141,3 +137,36 @@ def step(self, closure: Callable = None):
...
+            optimizer_conf.mutable_adam_conf().set_beta1(beta1)
+            optimizer_conf.mutable_adam_conf().set_beta2(beta2)
+            optimizer_conf.mutable_adam_conf().set_epsilon(epsilon)
+            optimizer_conf.mutable_adam_conf().set_do_bias_correction(
+                False
+            )  # TODO(zzk): Check this option
+
+            optimizer_conf.mutable_weight_decay_conf().set_weight_decay_rate(
+                weight_decay
+            )
```
Comment (strint) [inline python/oneflow/nn/optimizer/adamw.py:167]: 看了下lazy的实现，weight_decay的配置只需要这个一个配置就好

**Review by Ldpe2G — APPROVED** [review]
